internal/host/monitor.go

```diff
@@ -32,18 +32,25 @@ func (m *Manager) HostMonitoring() {
 		models.HostStatusInstallingInProgress,
 		models.HostStatusInstalled,
 	}
-
-	if err := m.db.Where("status IN (?)", monitorStates).Find(&hosts).Error; err != nil {
-		log.WithError(err).Errorf("failed to get hosts")
+	if err := m.db.Model(&models.Host{}).Where("status IN (?)", monitorStates).Count(&count).Error; err != nil {
+		log.WithError(err).Errorf("failed to count hosts")
 		return
 	}
-	for _, host := range hosts {
-		if !m.leaderElector.IsLeader() {
-			m.log.Debugf("Not a leader, exiting HostMonitoring")
+	for offset = 0; offset < count; offset += limit {
+		hosts := make([]*models.Host, 0, limit)
+		if err := m.db.Where("status IN (?)", monitorStates).Offset(offset).Limit(limit).
+			Order("cluster_id, id").Find(&hosts).Error; err != nil {
+			log.WithError(err).Errorf("failed to get hosts")
 			return
 		}
-		if err := m.RefreshStatus(ctx, host, m.db); err != nil {
-			log.WithError(err).Errorf("failed to refresh host %s state", *host.ID)
+		for _, host := range hosts {
+			if !m.leaderElector.IsLeader() {
+				m.log.Debugf("Not a leader, exiting HostMonitoring")
+				return
+			}
```

Comment:
Worth adding at the start of the method too?

Reply:
There is, it's the first line in the function.